File: android/app/src/main/java/com/example/audiomodem/protocol/Frame.kt
package com.example.audiomodem.protocol

import com.example.audiomodem.fec.CRC32

/**
 * Protocol frame encoding/decoding.
 * Format: [Type(1B)][SeqNum(1B)][PayloadLen(2B)][Payload][CRC-32(4B)]
 */
object FrameType {
    const val DATA: Byte = 0x01
    const val ACK: Byte = 0x02
    const val NACK: Byte = 0x03
    const val CONTROL: Byte = 0x04
    const val FILE_META: Byte = 0x05
    const val FILE_END: Byte = 0x06
    const val PING: Byte = 0x07
    const val PONG: Byte = 0x08
}

const val HEADER_SIZE = 4
const val MAX_PAYLOAD_SIZE = 1024
const val CRC_SIZE = 4

data class Frame(
    val type: Byte,
    val seqNum: Byte,
    val payloadLen: Int,
    val payload: ByteArray?
) {
    fun typeName(): String = when (type) {
        FrameType.DATA -> "DATA"
        FrameType.ACK -> "ACK"
        FrameType.NACK -> "NACK"
        FrameType.CONTROL -> "CONTROL"
        FrameType.FILE_META -> "FILE_META"
        FrameType.FILE_END -> "FILE_END"
        FrameType.PING -> "PING"
        FrameType.PONG -> "PONG"
        else -> "UNKNOWN(0x${type.toString(16)})"
    }

    fun encode(): ByteArray {
        val totalLen = HEADER_SIZE + payloadLen + CRC_SIZE
        val buf = ByteArray(totalLen)

        buf[0] = type
        buf[1] = seqNum
        buf[2] = (payloadLen shr 8 and 0xFF).toByte()
        buf[3] = (payloadLen and 0xFF).toByte()

        if (payloadLen > 0 && payload != null) {
            payload.copyInto(buf, HEADER_SIZE, 0, payloadLen)
        }

        // CRC over header + payload
        val dataForCRC = buf.copyOf(HEADER_SIZE + payloadLen)
        val crc = CRC32.compute(dataForCRC)
        buf[totalLen - 4] = (crc shr 24 and 0xFF).toByte()
        buf[totalLen - 3] = (crc shr 16 and 0xFF).toByte()
        buf[totalLen - 2] = (crc shr 8 and 0xFF).toByte()
        buf[totalLen - 1] = (crc and 0xFF).toByte()

        return buf
    }

    companion object {
        fun decode(data: ByteArray): Frame {
            require(data.size >= HEADER_SIZE + CRC_SIZE) { "Frame too short: ${data.size}" }

            val type = data[0]
            val seqNum = data[1]
            val payloadLen = ((data[2].toInt() and 0xFF) shl 8) or (data[3].toInt() and 0xFF)

            val expectedLen = HEADER_SIZE + payloadLen + CRC_SIZE
            require(data.size >= expectedLen) { "Frame truncated: ${data.size} < $expectedLen" }

            // Verify CRC
            val dataForCRC = data.copyOf(HEADER_SIZE + payloadLen)
            val expectedCRC = ((data[expectedLen - 4].toLong() and 0xFF) shl 24) or
                              ((data[expectedLen - 3].toLong() and 0xFF) shl 16) or
                              ((data[expectedLen - 2].toLong() and 0xFF) shl 8) or
                              (data[expectedLen - 1].toLong() and 0xFF)
            val actualCRC = CRC32.compute(dataForCRC)

            require(expectedCRC == actualCRC) {
                "CRC mismatch: expected 0x${expectedCRC.toString(16)}, got 0x${actualCRC.toString(16)}"
            }

            val payload = if (payloadLen > 0) data.copyOfRange(HEADER_SIZE, HEADER_SIZE + payloadLen) else null

            return Frame(type, seqNum, payloadLen, payload)
        }

        fun dataFrame(seqNum: Byte, payload: ByteArray) =
            Frame(FrameType.DATA, seqNum, payload.size, payload)

        fun ackFrame(seqNum: Byte) = Frame(FrameType.ACK, seqNum, 0, null)
        fun nackFrame(seqNum: Byte) = Frame(FrameType.NACK, seqNum, 0, null)
        fun pingFrame() = Frame(FrameType.PING, 0, 0, null)
        fun pongFrame() = Frame(FrameType.PONG, 0, 0, null)

        fun fileMetaFrame(filename: String, size: Long, md5: String): Frame {
            val nameBytes = filename.toByteArray()
            val md5Bytes = md5.toByteArray()
            val payload = ByteArray(2 + nameBytes.size + 8 + 32)

            payload[0] = (nameBytes.size shr 8 and 0xFF).toByte()
            payload[1] = (nameBytes.size and 0xFF).toByte()
            nameBytes.copyInto(payload, 2)
            val off = 2 + nameBytes.size
            for (i in 0 until 8) payload[off + i] = (size shr ((7 - i) * 8) and 0xFF).toByte()
            md5Bytes.copyInto(payload, off + 8)

            return Frame(FrameType.FILE_META, 0, payload.size, payload)
        }

        fun fileEndFrame() = Frame(FrameType.FILE_END, 0, 0, null)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Frame) return false
        return type == other.type && seqNum == other.seqNum && payloadLen == other.payloadLen
    }

    override fun hashCode(): Int = type.hashCode() * 31 + seqNum.hashCode()
}

File: pc/internal/protocol/frame.go
package protocol

import (
	"encoding/binary"
	"fmt"

	"github.com/jeongseonghan/audio-modem/internal/fec"
)

// Frame types
const (
	TypeData     byte = 0x01
	TypeACK      byte = 0x02
	TypeNACK     byte = 0x03
	TypeControl  byte = 0x04
	TypeFileMeta byte = 0x05
	TypeFileEnd  byte = 0x06
	TypePing     byte = 0x07
	TypePong     byte = 0x08
)

// Frame size limits
const (
	HeaderSize     = 4
	MaxPayloadSize = 1024
	CRCSize        = 4
)

// Frame represents a protocol frame.
// Format: [Type(1B)][SeqNum(1B)][PayloadLen(2B)][Payload][CRC-32(4B)]
type Frame struct {
	Type       byte
	SeqNum     byte
	PayloadLen uint16
	Payload    []byte
}

// TypeName returns a human-readable name for the frame type.
func (f *Frame) TypeName() string {
	switch f.Type {
	case TypeData:
		return "DATA"
	case TypeACK:
		return "ACK"
	case TypeNACK:
		return "NACK"
	case TypeControl:
		return "CONTROL"
	case TypeFileMeta:
		return "FILE_META"
	case TypeFileEnd:
		return "FILE_END"
	case TypePing:
		return "PING"
	case TypePong:
		return "PONG"
	default:
		return fmt.Sprintf("UNKNOWN(0x%02x)", f.Type)
	}
}

// NewDataFrame creates a new DATA frame.
func NewDataFrame(seqNum byte, payload []byte) *Frame {
	return &Frame{
		Type:       TypeData,
		SeqNum:     seqNum,
		PayloadLen: uint16(len(payload)),
		Payload:    payload,
	}
}

// NewACKFrame creates a new ACK frame.
func NewACKFrame(seqNum byte) *Frame {
	return &Frame{
		Type:       TypeACK,
		SeqNum:     seqNum,
		PayloadLen: 0,
		Payload:    nil,
	}
}

// NewNACKFrame creates a new NACK frame.
func NewNACKFrame(seqNum byte) *Frame {
	return &Frame{
		Type:       TypeNACK,
		SeqNum:     seqNum,
		PayloadLen: 0,
		Payload:    nil,
	}
}

// NewControlFrame creates a new CONTROL frame.
func NewControlFrame(payload []byte) *Frame {
	return &Frame{
		Type:       TypeControl,
		SeqNum:     0,
		PayloadLen: uint16(len(payload)),
		Payload:    payload,
	}
}

// NewPingFrame creates a new PING frame.
func NewPingFrame() *Frame {
	return &Frame{
		Type:       TypePing,
		SeqNum:     0,
		PayloadLen: 0,
		Payload:    nil,
	}
}

// NewPongFrame creates a new PONG frame.
func NewPongFrame() *Frame {
	return &Frame{
		Type:       TypePong,
		SeqNum:     0,
		PayloadLen: 0,
		Payload:    nil,
	}
}

// Encode serializes the frame to bytes with CRC-32.
func (f *Frame) Encode() []byte {
	totalLen := HeaderSize + int(f.PayloadLen) + CRCSize
	buf := make([]byte, totalLen)

	// Header
	buf[0] = f.Type
	buf[1] = f.SeqNum
	binary.BigEndian.PutUint16(buf[2:4], f.PayloadLen)

	// Payload
	if f.PayloadLen > 0 {
		copy(buf[HeaderSize:], f.Payload[:f.PayloadLen])
	}

	// CRC-32 over header + payload
	dataForCRC := buf[:HeaderSize+int(f.PayloadLen)]
	checksum := fec.CRC32(dataForCRC)
	binary.BigEndian.PutUint32(buf[totalLen-CRCSize:], checksum)

	return buf
}

// DecodeFrame deserializes bytes into a Frame, verifying CRC-32.
func DecodeFrame(data []byte) (*Frame, error) {
	if len(data) < HeaderSize+CRCSize {
		return nil, fmt.Errorf("frame too short: %d bytes", len(data))
	}

	f := &Frame{
		Type:       data[0],
		SeqNum:     data[1],
		PayloadLen: binary.BigEndian.Uint16(data[2:4]),
	}

	expectedLen := HeaderSize + int(f.PayloadLen) + CRCSize
	if len(data) < expectedLen {
		return nil, fmt.Errorf("frame truncated: have %d, need %d", len(data), expectedLen)
	}

	// Verify CRC
	dataForCRC := data[:HeaderSize+int(f.PayloadLen)]
	expectedCRC := binary.BigEndian.Uint32(data[expectedLen-CRCSize : expectedLen])
	actualCRC := fec.CRC32(dataForCRC)

	if expectedCRC != actualCRC {
		return nil, fmt.Errorf("CRC mismatch: expected 0x%08x, got 0x%08x", expectedCRC, actualCRC)
	}

	// Extract payload
	if f.PayloadLen > 0 {
		f.Payload = make([]byte, f.PayloadLen)
		copy(f.Payload, data[HeaderSize:HeaderSize+int(f.PayloadLen)])
	}

	return f, nil
}

// FrameToBytes converts a frame to bytes suitable for OFDM modulation.
// Includes RS encoding for forward error correction.
func FrameToBytes(f *Frame, rsEncoder *fec.RSEncoder) ([]byte, error) {
	raw := f.Encode()

	if rsEncoder != nil {
		encoded, err := rsEncoder.Encode(raw)
		if err != nil {
			return nil, fmt.Errorf("RS encode: %w", err)
		}
		return encoded, nil
	}

	return raw, nil
}

// BytesToFrame decodes bytes from OFDM demodulation back to a frame.
// Includes RS decoding for error correction.
func BytesToFrame(data []byte, rsDecoder *fec.RSEncoder) (*Frame, error) {
	var decoded []byte

	if rsDecoder != nil {
		var err error
		decoded, err = rsDecoder.Decode(data)
		if err != nil {
			return nil, fmt.Errorf("RS decode: %w", err)
		}
	} else {
		decoded = data
	}

	return DecodeFrame(decoded)
}
